### ISSUE TITLE
alrpLoss not convergence

### ISSUE THREAD

**Issue by jinyl777** [kemaloksuz/aLRPLoss#7]
First of  all, it is extermely intersting work. I use it to my task for detect to pedestrain ， just like you open code but the alrpLoss not convergence ，what is more  the  regloss convergence normly.

**Comment by kemaloksuz** [comment]
Can you provide some extra information (e.g. some log file including your configuration)?

**Comment by jinyl777** [comment]
of course, but I train it in a other frame ， I use Resnet50 + FPN and generate anchor to classifiy the pedestrain or backgroud. Then
I use the same input like your exeample , Just started the loss_box * 50 is about 0.2 .but when the SB.wright updata after a epoch.
the loss is about 1.2 and after I trian it 100 epoch the loss is always about 1.2. What's more the same time regloss  is  normly from 2 to 0.4. Can you give me some advice to deal with this problem

**Comment by kemaloksuz** [comment]
It is a bit difficult to validate your implementation but I can try: What about the classification loss? Does it decrease regularly?  Can you please send me a log file/plot that shows how classificatoin and regression losses change during training?

**Comment by jinyl777** [comment]
[run_2021_01_09_18_41_28.log](https://github.com/kemaloksuz/aLRPLoss/files/5792038/run_2021_01_09_18_41_28.log)
Ihis is the log file. you can see  the different between calssification with alrp

**Comment by kemaloksuz** [comment]
What is "Cls loss" in the log file? I think it is cross entropy or focal loss, not aLRP Classification component?

I can recommend you to debug your implementation by checking the following:

- First ensure that the following variable is set to "number of iterations per epoch":
https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L59

- Then please set the following variable to 1 to see unweighted/raw aLRP classification and aLRP regression losses:
https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L57 

- Now you can log the following two values. In the first one, you should check "losses_cls" variable, in the second line that I provide please check "losses_bbox":
https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L219

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L238

- Up to here, self balancing mechanism is cancelled in the first epoch to observe loss values. Now, we should expect (i) losses_cls to start from 1 in the beginning and decrease; (ii) losses_bbox to start from 0 and increase. We optimize a single loss, which is their summation, so (iii) "losses_cls+losses_bbox" should also decrease. This summation is exactly the aLRP value, and the only loss. Now all three loss values (losses_cls, losses_bbox or their summation) should be between 0 and 1 (there may be tolerable violations only in the first iterations due to precision), else there is a problem in the code. Maybe you can run one epoch and log/print these three losses every iteration.  It is better to check aLRP classification and aLRP regression seperately to see if there is some error.

- If you are sure up to here, I would suggest three more checks concerning the inputs of aLRP Loss in the following line:

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L219

flat_preds should be a vector captures unnormalized outputs obtained before sigmoid.
flat_labels should be a vector for labels of flat_preds. If an example is positive, negative or ignored; the corresponding element is 1, 0 and -1  respectively.
loss_bbox is a vector containing the IoU/GIoU Losses of only positive examples. The orders of the losses should follow the positive examples in flat_preds and flat_labels, such that all three inputs match.  

- If you validate all these points, you can reactivate self balance. If you still you have a convergence issue, then you can search a learning rate or review how you combine aLRPLoss with other losses (if any, I see some classification loss and reid loss in the log). 

Hope this helps.

**Comment by jinyl777** [comment]
Thanks very much I test my code and after train two hours  I find the classfication_loss is about 0.99  and never down ,and the boxes_loss is about more than  0.1 the sum is more than 1 . So think I got the problem where has occured .
